core/src/test/java/org/apache/iceberg/hadoop/TestCachingCatalog.java

```diff
@@ -257,6 +264,44 @@ public void testCacheExpirationEagerlyRemovesMetadataTables() throws IOException
             catalog.cache().asMap().containsKey(metadataTable)));
   }
 
+  @Test
+  @Ignore("reproduces https://github.com/apache/iceberg/issues/3791")
+  public void testDeadlock() throws IOException, InterruptedException {
+    HadoopCatalog underlyingCatalog = hadoopCatalog();
+    TestableCachingCatalog catalog = TestableCachingCatalog.wrap(underlyingCatalog, Duration.ofSeconds(1), ticker);
+    Namespace namespace = Namespace.of("db", "ns1", "ns2");
+    int numThreads = 20;
```

Comment:
Would it make sense to use fewer threads, so the space for collision / hitting the deadlock is smaller? And then iterating like you are a few extra times? This way the `Random.nextInt` calls are more likely to collide and we don't have to spawn so many threads.

Reply:
Can you hack the hashCode (e.g. make constant) to coerce collisions?

Reply:
I don't mean to take up too much of your time, but you mean for the cache key's (the TableIdentifier)?

Would it be equivalent to just make one cache entry and then just only operate on that? That would make dropping the table (cache entry) free.

Reply:
Yes, the key. The map locks on a hashbin so locking multiple in an unpredictable order is a classic deadlock case. I don’t know if a single entry would suffice or if you need threads performing ABA vs BAB to get the desired test case.

Reply:
The thing here is to make the cache miss (from my tests). I changed this a little to get rid of the random stuff, making cache misses more certain.